### ISSUE TITLE
diff coloring not working

### ISSUE THREAD

**Issue by 1268** [martinvonz/jj#1930]
## Description

for some reason JJ is not coloring the added lines with a diff:

![jj](https://github.com/martinvonz/jj/assets/126826528/06f14a9e-9e9a-466a-9aad-0a13fcf4bb46)

compare to Git:

![git](https://github.com/martinvonz/jj/assets/126826528/fa2b1e59-a329-48a6-a227-fba8e5b584a8)

## Specifications

<https://github.com/martinvonz/jj/releases/download/v0.8.0/jj-v0.8.0-x86_64-pc-windows-msvc.zip>

**Comment by martinvonz** [comment]
Interesting that the first few lines are cyan and the next line is highlighted. Maybe we need to use different colors codes on Windows. Which terminal is that?

**Comment by 1268** [comment]
https://github.com/microsoft/terminal

**Comment by 1268** [comment]
https://github.com/microsoft/terminal

looks like JJ is doing this currently:

~~~
\x1b[38;5;2m+asdf\x1b[39m
~~~

and Git is doing this:

~~~
\x1b[32m+\x1b[m\x1b[32masdf\x1b[m
~~~

**Comment by 1268** [comment]
I think its something specific to JJ. using this Go program:

~~~go
package main

import "fmt"

func main() {
   fmt.Println("\x1b[38;5;2m+asdf\x1b[39m")
   fmt.Println("\x1b[32m+\x1b[m\x1b[32masdf\x1b[m")
}
~~~

I get expected result:

![pass](https://github.com/martinvonz/jj/assets/126826528/452f6034-c5c3-4ccf-a266-bf016cd6baac)

it might be one of the other previous sequences thats breaking it, not sure. I will try to check more when I get home.

**Comment by martinvonz** [comment]
Well, that's good news :)

**Comment by 1268** [comment]
note for myself work computer is Windows Terminal 1.17.11461

**Comment by 1268** [comment]
made some progress. turns out its the shell, not the terminal. currently I use this:

~~~
> $PSVersionTable
PSVersion                      7.3.0
PSEdition                      Core
~~~

which fails as originally reported, but for some reason if I switch to the old system version:

~~~
> $PSVersionTable
PSVersion                      5.1.18362.1110
PSEdition                      Desktop
~~~

I get correct colors. will keep checking. if you dont mind please keep issue open for now until I nail this down.

**Comment by 1268** [comment]
OK narrowed it down more, its NOT actually the terminal OR shell. it seems JJ is trying to spawn `less.exe`. if it does NOT find it, then everything works correct. if it DOES find it, then it uses it and the colors fuck up - so maybe Less has some configuration I need to change? not sure, will keep checking

**Comment by martinvonz** [comment]
Probably useful: https://github.com/martinvonz/jj/blob/5bd3dca21b56fa4d1c61c60288974d6143580be6/src/config/misc.toml#L8

**Comment by 1268** [comment]
yeah, it seems Less is the culprit:

![less](https://github.com/martinvonz/jj/assets/126826528/39a371a5-d0db-42bc-b435-06f115eab53d)

I am not sure how to fix that.

~~~
> less --version
less 633 (Spencer V8 regular expressions)
~~~

I will keep researching.

**Comment by 1268** [comment]
https://github.com/jftuga/less-Windows/issues/40

**Comment by 1268** [comment]
OK so normally I would say close this issue, but until Less is fixed, would you consider some option to use 3/4 bit color instead of 8-bit? thats what Git proper uses. if not I understand

**Comment by martinvonz** [comment]
Sure, I'm not opposed to that, but I won't have time to implement it myself. Are you interested in doing that? Do you know how we would determine whether to use 16 or 256 colors? I don't know if git detects it somehow or if it always uses 16 colors (codes 30-37 + 1 for bold/bright) on Windows.

**Comment by 1268** [comment]
OK I just noticed, if you do lower case r instead, it works:

![fix](https://github.com/martinvonz/jj/assets/126826528/777c7a75-9295-471a-8413-1ba000d26871)

but you have to hit y to proceed. maybe some flag is available that bypasses the question, not sure yet. but that might be an option.

**Comment by 1268** [comment]
https://github.com/gwsw/less/issues/410

**Comment by ilyagr** [comment]
I'm pretty sure we already use only 16 colors. It seems like `\x1b[38;5;2m` is an instruction [from the 256-color](https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit), but the color it uses is green.

I'd suggest using 16 colors forever, since they adjust to the terminal theme. I guess the solution would be to use ``\x1b[32m` from the [4-bit world](https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit) instead of `\x1b[38;5;2m`.

**Comment by 1268** [comment]
I would agree with the previous comment. unless JJ is actually using colors outside of the standard 16, all the current code is doing is hurting backward compat for no benefit.

**Comment by martinvonz** [comment]
Commit 5cf2b6615a7aa3d5607139605b6a78bd9b16320f has some discussion about the switch from 8 colors and using bold for bright (code 1). With the switch to 16 actual colors, we can set boldness and bright colors independently.

**Comment by 1268** [comment]
OK well since Less apparently sucks, I switched to Slit for now:

https://github.com/tigrawap/slit

I created this file:

~~~
C:\Users\Steven\.jjconfig.toml
~~~

with this content:

~~~toml
[ui.pager]
command = ["slit"]
~~~

and now it just works:

![slit](https://github.com/martinvonz/jj/assets/126826528/07e6ad2d-e923-452f-b7e5-9c0d34d1c8be)

I guess issue could probably be closed, but I will leave open in case you want to document this issue, or maybe just wait to see if its resolved upstream.

**Comment by ob** [comment]
I feel like there's a deeper issue here. I'm on macOS and if I run (building `main` branch from GitHub source):

```
jj ❯ PAGER='less -XG' cargo run -- -h
```

I get uninterpreted escapes. 

<img width="688" alt="Screenshot 2023-08-01 at 12 50 23 PM" src="https://github.com/martinvonz/jj/assets/4950/acd00e1e-22d7-4b7c-b896-ba212340989b">


But I run:

```
jj ❯ cargo run -- --color=always -h | less -XG
```

I get proper colors. 

<img width="888" alt="Screenshot 2023-08-01 at 12 50 35 PM" src="https://github.com/martinvonz/jj/assets/4950/a3b17e4f-3284-4ed6-a12e-5fdf8f2b2482">

**Comment by 1268** [comment]
I dont think thats the same issue. you should be running `less -R` regardless, and in both cases you are not.

**Comment by 1268** [comment]
https://github.com/gwsw/less/issues/377

**Comment by 1268** [comment]
update, looks like on Windows `-D a` (or `-Da` needs to be used with any Less command, in order to support 8-bit color. 

@martinvonz would you consider modifying the default Less options? or will users just need to configure that on their own?

**Comment by 1268** [comment]
@martinvonz upon reading more comments in that linked thread it seems the issue is not fully resolved. I would say for now the best action for Windows users is to avoid Less, and switch to another pager. However for Windows users insisting on Less, they can use:

~~~
C:\Users\Steven\.jjconfig.toml
~~~

with something like:

~~~
[ui.pager]
command = ["less", "-D", "a"]
~~~

**Comment by martinvonz** [comment]
Sorry I didn't get back to you earlier. Since we seem to have very few Windows users and very few people with even access to a Windows computer, I think we should just trust your judgement :) So feel free to send a PR splitting out `ui.pager` into per-platform files. Thanks.

Even if it turns out that there's a better choice than `less -D a`, it seems clear that we'll eventually want per-platform defaults, so it should be a step in the right direction.

**Comment by 1268** [comment]
closing this. I would recommend Windows users switch to the Slit pager or some other, as Less just seems to be outdated at this point, and they even closed the issue that I reported. see my previous comment if you need a link to Slit

**Comment by 3052** [comment]
just to add, it seems that this has been fixed with version v0.9.0 (Sep 6, 2023) and later, even with the default config.
